src/classes/PMT_Payment_TDTM.cls

```diff
@@ -201,4 +162,132 @@ public class PMT_Payment_TDTM extends TDTM_Runnable {
+
+        return new Map<Id, Opportunity>((List<Opportunity>)Database.query(soql));
+    }
+
+    /*******************************************************************************************************
+    * @description          For multicurrency orgs, validate the currencies and set the appropriate stages
+    *                       and currency fields
+    * @param newlist        The list of Payments from trigger new.
+    * @param oldlist        The list of Payments from trigger old.
+    * @param triggerAction  Which trigger event (BeforeInsert, AfterInsert, etc.).
+    * @return dmlWrapper    This method modifies dmlWrapper and returns it
+    ********************************************************************************************************/
+    private DmlWrapper handleMultiCurrencyPayments(
+        List<npe01__OppPayment__c> newlist, 
+        List<npe01__OppPayment__c> oldlist,
+        TDTM_Runnable.Action triggerAction, 
+        DmlWrapper dmlWrapper) {
+
+        if (dmlWrapper == null) {
+            dmlWrapper = new DmlWrapper();
+        }
+
+        Set<Id> oppIds = new Set<Id>();
+        for (npe01__OppPayment__c pmt : newlist) {
+            oppIds.add(pmt.npe01__Opportunity__c);
+        }
+
+        Map<Id, Opportunity> mapOfOpps = getPaymentOpportunities(oppIds);
+        
+        // Validation and default the CurrencyIsoCode for new records for Before Insert/update
+        if (isCurrencyFieldAvailable && (triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate)) {
+            for (Integer i = 0; i < newlist.size(); i++) {
+                npe01__OppPayment__c pmt = newlist[i];
+                String oldPmtCurrency = (oldList != null ? (String) oldlist[i].get('CurrencyIsoCode') : null);
+                setPaymentCurrency(mapOfOpps, pmt, oldPmtCurrency, triggerAction);
+            }
+        }
+
+        // For an After Insert/Update Payment action, when Advanced Curr Mgt is enabled and there is a defined
+        // Closed stage name in settings, roll up the total paynents and compare to the Opp.Amount. If the
+        // two values are equal then close the Opportunity by setting the stage name.
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert || triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+
+            npe01__Contacts_And_Orgs_Settings__c ContactSettings = UTIL_CustomSettingsFacade.getContactsSettings();
+            String closedStageName = ContactSettings.Payments_Auto_Close_Stage_Name__c;
+
+            Boolean isACMEnabled = UTIL_Currency.getInstance().isAdvancedCurrencyManagementEnabled();
+            if (isACMEnabled && String.isNotEmpty(closedStageName)) {
+                // Loop through all the queried Opps with the subquery on payments
+                for (Opportunity opp : mapOfOpps.values()) {
+                   handleAdvancedCurrencyPayments(opp, closedStageName, dmlWrapper);
```

Comment:
Maybe this should be `handleOppPaymentWithAdvancedCurrency()` since it operates on a single Opp at a time.